### PR TITLE
Web UI uses Webdav instead of ajax/* calls

### LINKED ISSUE #12353: Move files app to use WebDAV instead of custom PHP endpoints
Currently we have at least two endpoints to manage files:
1) WebDAV
2) Private PHP ajax endpoints in the files app

To reduce code duplication and endpoint duplication one idea would be to make the files app use the WebDAV endpoint instead, just like any other external client does (mobile, desktop, etc)

An existing JS library could be used (for example: https://github.com/sara-nl/js-webdav-client) by the files app.

We need to make sure that every custom property (ex: "icon", "isPreviewAvailabe", etc) is available on the WebDAV endpoint to keep the existing functionality working.

There might be other challenges like batch file deletion, where the JS code would need to send multiple "DELETE" calls, one for every selected file, instead of deleting a series of files in one call.

@DeepDiver1975 @icewind1991

### PR DESCRIPTION
**Pull request by PVince81**
Fixes https://github.com/owncloud/core/issues/12353
- [x] list
- [x] move
- [x] rename
- [x] new file
- [x] new folder
- [x] delete
- [x] download
- [x] delete legacy ajax code
- [x] fix jsocclient namespace / extract as separate lib ?
- [x] merge OC.Files.FileInfo into OCA.Files.FileInfo
- [x] remove callbacks, only use Promises
- [x] remove "isPreviewAvailable"
- [x] isSharedMount ?
- [x] error cases, forward status
- [x] older error messages ? (might be missing from Webdav)
- [x] check missing properties (isPreviewAvailable, etc)
- [x] build and verify JS docs!
- [x] unit tests for the new JS classes
- [ ] TODO separately ?
  - [ ] displayed download links in webdav format (public and internal)
### Tests
- [x] TEST: file picker (also uses list.php)
- [x] TEST: public page
- [x] TEST: favorites list still works + file operations there
- [x] TEST: sharing lists still work + file operations there
- [x] TEST: trashbin still works + file operations
- [x] TEST: storage not available exception
- [x] TEST: IMPORTANT: test in IE8 !!! 
- [x] TEST: external storage section in sidebar
##### Move
- [x] TEST: move multiple files
- [x] TEST: move all files
- [x] TEST: move file where source does not exist
- [x] TEST: move file where target does not exist
- [x] TEST: move file where target already exists
- [x] TEST: move file where source folder has been removed
##### Rename
- [x] TEST: simple rename
- [x] TEST: rename file where source does not exist any more
- [x] TEST: rename file where target already exists
- [x] TEST: rename file where source folder has been removed
- [x] TEST: rename file with invalid chars
##### Create file
- [x] TEST: create file in current folder
- [x] TEST: create file when current folder does not exist any more
- [x] TEST: create file when file already exists => shows error message
##### Create folder
- [x] TEST: create folder
- [x] TEST: create folder when current folder does not exist any more
- [x] TEST: create folder when folder already exists
##### Delete
- [x] TEST: delete single file
- [x] TEST: delete multiple files
- [x] TEST: delete all files
- [x] TEST: delete file when source folder does not exist any more
- [x] TEST: delete file that does not exist any more
##### Download
- [x] TEST: single file download (webdav URL)
- [x] TEST: multiple files download (ajax URL for now, zip)
- [x] TEST: folder download (ajax URL for now, zip)
- [x] TEST: single file public page download
- [x] TEST: multiple files public page download
- [x] TEST: folder download public page download
- [x] TEST: public page, download whole page (click download on the top right)
##### Other
- [x] TEST: timestamps and timezone, to make sure the library's parsing works as intended
- [x] TEST: favorites appear in the regular file list
- [x] TEST: share owner visible in incoming shares
- [x] TEST: various permissions for files/folders
- [x] TEST: previews and mime icons still appear
- [x] TEST: single file preview in public page
- [x] TEST: MOVE/COPY in reverse proxy environments, because the "Destination" header is absolute, we need to be sure it points to the correct URL

### CODE REVIEW

**Comment by karlitschek** [comment]
Wow. Very cool!!

File: core/js/jsocclient.js [hunk 1]
```diff
@@ -0,0 +1,512 @@
...
+		 *
+		 * @param {String} path path to retrieve
+		 * @param {Function} callback callback that receives an array
+		 * of files as parameter
+		 * @param {boolean} [includeParent=false] set to true to keep
+		 * the parent folder in the result list
+		 *
+		 * @return {Promise} promise 
+		 */
+		list: function(path, callback, includeParent) {
```
Comment (icewind1991) [inline core/js/jsocclient.js]: why add a callback when we're already returning a promise?

Comment (PVince81) [inline core/js/jsocclient.js]: Good point. I was more thinking of people not familiar with promises.

Also I'm not sure yet whether I want this "lib" to depend on jquery at all (which provides the promises implementation).

I'll add to my TODO list to reconsider this. Thanks.

**Comment by PVince81** [comment]
- [ ] TODO: redundant callback/promise

**Comment by rullzer** [comment]
Awesome! Stuff like this makes me sorry I left my laptop at home. ;-) 

On June 12, 2015 1:40:52 PM GMT+02:00, Vincent Petry <notifications@github.com> wrote:
>Fixes https://github.com/owncloud/core/issues/12353
>
>
>- [x] list
>- [x] move
>- [x] rename
>- [x] new file
>- [x] new folder
>- [x] delete
>- [ ] upload
>- [ ] upload with fallback ?
>- [x] download
>- [ ] download (use downloadURL ?)
>- [ ] multi-download
>
>- [ ] delete legacy code (list.php, etc)
>
>- [ ] file picker (also uses list.php)
>- [ ] test: storage not available exception
>
>- [ ] file ids
>
>- [ ] multi-file selection broken?
>
>- [ ] error cases, forward status
>- [ ] older error messages ? (might be missing from Webdav)
>
>- [ ] refresh UI with infos like mimetype, etc
>
>- [ ] token/reload logic (make this more global!)
>
>- [ ] merge OC.Files.FileInfo into OCA.Files.FileInfo
>- [ ] check missing properties
>
>- [ ] fix jsocclient namespace / extract as separate lib ?
>
>- [ ] leverage etag knowledge and use preconditions ?
>
>- [ ] missing previews: use @rullzer's mimetype list code ?
>- [ ] remove "isPreviewAvailable"
>- [ ] isSharedMount ?
>
>- [ ] share properties from sharing ?
>- [ ] external storage properties ?
>
>- [ ] allow passing extra properties to list() calls (required for
>favorites, etc)
>- [ ] favorites
>
>- [ ] webdav download: x-sendfile!!
>
>- [ ] public page list
>- [ ] trashbin broken => split FileList into AbstractFileList (base
>class without) and HomeFileList (for "All files" specific code, using
>WebDAV)
>- [ ] check other sidebar sections
>
>- [ ] IMPORTANT: test in IE8 !!! might require IE8 shim
>
>- [ ] build and verify JS docs!
>
>- [ ] sort ? (now client side)
>
>- [ ] getstoragestats: make it use root info ?
>
>- [ ] batch multiple webdav into a single request?
>
>- [ ] JS unit tests
>
>
>You can view, comment on, or merge this pull request online at:
>
>  https://github.com/owncloud/core/pull/16902
>
>-- Commit Summary --
>
>  * Add js-webdav-client library
>  * Added class to manage files through webdav
>  * Added more webdav ops
>
>-- File Changes --
>
>    D apps/files/ajax/delete.php (83)
>    D apps/files/ajax/move.php (60)
>    D apps/files/ajax/newfile.php (104)
>    D apps/files/ajax/newfolder.php (100)
>    D apps/files/ajax/rename.php (57)
>    M apps/files/js/file-upload.js (44)
>    M apps/files/js/filelist.js (138)
>    M apps/files/js/files.js (19)
>    M apps/files/lib/app.php (88)
>    M bower.json (5)
>    M core/js/js.js (8)
>    A core/js/jsocclient.js (512)
>    A core/vendor/js-webdav-client/.bower.json (15)
>    A core/vendor/js-webdav-client/.gitmodules (3)
>    A core/vendor/js-webdav-client/COPYING.LESSER.txt (165)
>    A core/vendor/js-webdav-client/LICENSE.txt (674)
>    A core/vendor/js-webdav-client/scripts/check_dependencies.sh (25)
>    A core/vendor/js-webdav-client/tests/resources/mock.js (468)
>    A core/vendor/js-webdav-client/tests/resources/qunit.css (244)
>    A core/vendor/js-webdav-client/tests/resources/qunit.js (2212)
>   A core/vendor/js-webdav-client/tests/templates/tests_footer.html (2)
>  A core/vendor/js-webdav-client/tests/templates/tests_header.html (12)
>  A core/vendor/js-webdav-client/tests/unittests/0000_exception.js (46)
>    A core/vendor/js-webdav-client/tests/unittests/0010_codec.js (72)
>  A core/vendor/js-webdav-client/tests/unittests/0020_property.js (175)
>  A core/vendor/js-webdav-client/tests/unittests/0030_response.js (153)
>A core/vendor/js-webdav-client/tests/unittests/0040_multistatus.js (72)
>   A core/vendor/js-webdav-client/tests/unittests/0050_client.js (1065)
>A
>core/vendor/js-webdav-client/tests/unittests/plugins/acl/0000_privilege.js
>(104)
>A core/vendor/js-webdav-client/tests/unittests/plugins/acl/0010_ace.js
>(159)
>A core/vendor/js-webdav-client/tests/unittests/plugins/acl/0020_acl.js
>(186)
>A
>core/vendor/js-webdav-client/tests/unittests/plugins/acl/0030_acl_codec.js
>(179)
>A
>core/vendor/js-webdav-client/tests/unittests/plugins/acl/0040_client_acl.js
>(223)
>A
>core/vendor/js-webdav-client/tests/unittests/plugins/acl/current-user-privilege-set_codec.js
>(75)
>A
>core/vendor/js-webdav-client/tests/unittests/plugins/acl/inherited-acl-set_codec.js
>(66)
>A
>core/vendor/js-webdav-client/tests/unittests/plugins/acl/principal-collection-set_codec.js
>(66)
>A
>core/vendor/js-webdav-client/tests/unittests/plugins/webdav_default_codecs/creationdate.js
>(53)
>A
>core/vendor/js-webdav-client/tests/unittests/plugins/webdav_default_codecs/getcontentlength.js
>(53)
>A
>core/vendor/js-webdav-client/tests/unittests/plugins/webdav_default_codecs/getlastmodified.js
>(54)
>A
>core/vendor/js-webdav-client/tests/unittests/plugins/webdav_default_codecs/owner.js
>(89)
>A
>core/vendor/js-webdav-client/tests/unittests/plugins/webdav_default_codecs/resourcetype.js
>(72)
>    M lib/base.php (3)
>
>-- Patch Links --
>
>https://github.com/owncloud/core/pull/16902.patch
>https://github.com/owncloud/core/pull/16902.diff
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/owncloud/core/pull/16902

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by PVince81** [comment]
@rullzer coding on the beach is probably not a good idea anyway, you don't want sand in your laptop.

**Comment by DeepDiver1975** [comment]
@PVince81 rebase please

Can we do this step by step? As in merge now and fix remaining issues in a followup prs?

**Comment by PVince81** [comment]
Yeah, I'll try to split them to avoid causing regressions.

**Comment by DeepDiver1975** [comment]
> Yeah, I'll try to split them to avoid causing regressions.

I guess there is not need to split the existing pr - just bring this to a working state and merge ....

**Comment by PVince81** [comment]
Working state is a big word. There is still quite some work needed because PROPFIND doesn't return all the information we used to have with the ajax call so there is a high risk of regression.

By splitting I could remove the "list" operation and at least merge the simpler ones.

I'll see what I can do.

**Comment by DeepDiver1975** [comment]
> I'll see what I can do.

thx

**Comment by PVince81** [comment]
Rebased

**Comment by PVince81** [comment]
- [ ] BUG: public page broken

I'll move the todos back to the original ticket to be done separately.
Trying to get a working version from what we already have here :smile: (working as in "does not break other things")

**Comment by PVince81** [comment]
@icewind1991 ok I think I'll only use promises, no callbacks.
This means the client lib will currently rely on jQuery, but that's fine for now.

**Comment by icewind1991** [comment]
> @icewind1991 ok I think I'll only use promises, no callbacks.
This means the client lib will currently rely on jQuery, but that's fine for now.

I wish we could just use babel and write es6/7, the `async`/`await` syntax is soooo nice :smile:

**Comment by PVince81** [comment]
Yeah, but we'd need an IE8 shim or something. Maybe one day...

**Comment by PVince81** [comment]
- [ ] BUG: creating a text file can overwrite existing one (need to find precondition or a way to detect if it exists)

**Comment by PVince81** [comment]
Public page now works and also supports rename, delete, move, create folder :smile:

**Comment by PVince81** [comment]
- [ ] TODO: auth error cases in reloadCallback (token expired, etc)

**Comment by PVince81** [comment]
- [ ] debounce operations ? for example when deleting a lot of files, a DELETE call is made for each file (maybe the browser already does that)

**Comment by PVince81** [comment]
Added better error handling and failing promises.
It is nice to see that Webdav has specific codes to distinguish "target already exists" or "parent folder does not exist for the resource you wanted to touch", etc. Makes it possible to display customized error messages.

**Comment by PVince81** [comment]
- [ ] missing share owner info in response (indicator missing): new Webdav property ?
- [ ] missing external storage icon: could be done as an additional: new Webdav property ? or reuse resource type ?

**Comment by DeepDiver1975** [comment]
> missing external storage icon: could be done as an additional: new Webdav property ? or reuse resource type ?

our permissions property already contains M - to tell if a folder/file is mounted

**Comment by PVince81** [comment]
Right, I could use that. But the old web UI had more detailed info about the mount type.
Also at some point we'll want to use the specific icon for ext storage mounts (WND, etc).

**Comment by PVince81** [comment]
- [ ] TEST: various permissions for files/folders

**Comment by PVince81** [comment]
Solved the mount types without extra attributes.

But:
- [ ] set "data-share-owner" attribute required to show owner, will require a new Webdav attribute

**Comment by PVince81** [comment]
- [ ] abort ajax call: does not seem possible with promises

**Comment by PVince81** [comment]
- [ ] rewire existing unit tests to check if OC.Files.Client is called properly (stub it)
- [ ] write tests for OC.Files.Client and check with sinon's fakeServer that the requests are made properly
- [ ] also add convenience methods on OC.Files.Client, even if they still need ajax endpoint ? (ex: downloading multiple files)

**Comment by PVince81** [comment]
- [ ] make publicpreview work when s2s is disabled (maybe use a special header)

**Comment by PVince81** [comment]
@oparoz @rullzer regarding previews: if I remember well there was a decision to simply let them fail with 404 when they don't exist ?
My problem now is that the Webdav response doesn't contain "isPreviewAvailable" any more, so I'd need to indiscriminately request the preview for any file (except folders). Probably at some point we'll also wait to have a `OC.Previews` namespace with a list of supported previews.

Ideally I'd like to get rid of "isPreviewAvailable" now. It makes the logic quite complex.

**Comment by rullzer** [comment]
@PVince81 the OC.Previews namespace sounds like a good approach... makes it also more clean IMO. Just drop the isPreviewAvailable.

**Comment by PVince81** [comment]
Ok thanks @rullzer, raised here https://github.com/owncloud/core/issues/17543

**Comment by PVince81** [comment]
Without isPreviewAvailable, we might get a lot of 404 for files that don't have previews. In this case we should look into caching the 404 errors: https://github.com/owncloud/core/issues/17557

**Comment by PVince81** [comment]
- [ ] remove "OC.files" convenience helper, introduce "OC.getFilesClient()" instead ?

**Comment by PVince81** [comment]
I split some commits as separate PRs:
- [ ] REQUIRES: sinon lib update https://github.com/owncloud/core/pull/17555
- [ ] REQUIRES: OC.joinPaths addition: https://github.com/owncloud/core/issues/17559

**Comment by PVince81** [comment]
- [ ] TEST: spaces and unicode characters

**Comment by PVince81** [comment]
- [ ] TEST: timestamps and timezone, to make sure the library's parsing works as intended

**Comment by PVince81** [comment]
So, I added unit tests for the client class.
There's a few more missing regarding the twisted permissions parsing.

**Comment by PVince81** [comment]
- [ ] REQUIRES https://github.com/owncloud/core/pull/17601: allowing public webdav access even when outgoing remote shares are disabled, do this by detecting ajax requests

**Comment by DeepDiver1975** [comment]
@PVince81 a separate job for this pr: https://ci.owncloud.org/job/server-master-linux-jsocclient/

**Comment by PVince81** [comment]
- public webdav should not return mount info, this will influence icons too, raised here https://github.com/owncloud/core/pull/17605

**Comment by PVince81** [comment]
- [ ] REQUIRES: OC functions to get host, port, etc for some unit tests https://github.com/owncloud/core/pull/17609

**Comment by PVince81** [comment]
- [ ] BUG: browser shows basic auth prompt dialog when session expired / server returns 401

**Comment by PVince81** [comment]
Of course, I forgot to remove this:
```
PHP Fatal error:  Call to undefined method OCA\Files\App::rename() in /var/jenkins/workspace/server-master-linux-jsocclient/database/sqlite/label/SLAVE/apps/files/tests/ajax_rename.php on line 109
```

**Comment by PVince81** [comment]
- [ ] REQUIRES: OC.joinPaths fix https://github.com/owncloud/core/pull/17599

**Comment by PVince81** [comment]
Going to squash this and then split again into sensible (revertable) commits.

**Comment by PVince81** [comment]
- [ ] BUG: public link: drag file into subdir doesn't properly update the view

**Comment by PVince81** [comment]
Squashed to fewer commits that show the progression.

**Comment by PVince81** [comment]
- [ ] missing "favorite" info

**Comment by PVince81** [comment]
I moved the remaining todos together here: https://github.com/owncloud/core/pull/16902#issuecomment-120999181
And added a few tests in the original post: https://github.com/owncloud/core/pull/16902#issue-87703101

Getting close, but still not 100% there.
And we're still missing Webdav upload which I'll do separately later.

**Comment by rullzer** [comment]
@PVince81 I'll (finally) have a look later this week. But I just wanted to say awesome work!

**Comment by PVince81** [comment]
And hell begins... doesn't work in IE8. es5-shim doesn't seem to help.
The library we use requires Object.defineProperty.

**Comment by MorrisJobke** [comment]
> And hell begins... doesn't work in IE8. es5-shim doesn't seem to help.
> The library we use requires Object.defineProperty.

https://github.com/owncloud/core/issues/15567#issuecomment-121149990

**Comment by PVince81** [comment]
- [ ] BUG: library uses JS APIs not supported and not shimmable in IE8
- [ ] BUG: IE9 shows folders as files

**Comment by PVince81** [comment]
I have a fix for IE9 that I will push soon.

Regarding IE8 I did some research and found out that:
- the library we use uses the following methods not supported in IE8:
    - Object.defineProperty
    - document.implementation.createDocument (to create an XML document)
    - node.createElementNS (to create a DOM node with namespace)
- the following shims are available:
    - https://github.com/es-shims/es5-shim for Object.defineProperty
    - https://gist.github.com/brettz9/4274369 for document.implementation.createDocument
    - Still looking for a shim for node.createElementNS
- there aren't many JS libs out there that does Webdav, and the few ones that does assume a modern browser (aka not IE8)

And if we have to load 3-4 other libraries just to make this work on IE8. Also these libraries might cause other side effects even in non-IE8 envs, unless we limit them to IE8 env.

I have the feeling that it might not be worth the hassle. This means that this PR will have to wait for IE8 to be unsupported (https://github.com/owncloud/core/issues/15567#issuecomment-121193211).

IE9 seems to work fine though.

**Comment by PVince81** [comment]
TIL that one can use `If-Match-None: *` with PUT to tell the server "only put this file if there is no file that already exists". It then returns 412 Precondition Failed if the file already exists.
That HTTP stuff is neat :smile:

**Comment by PVince81** [comment]
I tried to replace the sara webav client with davclient.js in this POC: https://github.com/owncloud/core/pull/17668
But had to heavily modify davclient.js to work with IE8. Scary stuff.

**Comment by PVince81** [comment]
Rebased onto master.

- [ ] TODO: fix unit tests again
- [ ] TODO: bring in the IE8 fixes from https://github.com/owncloud/core/issues/17668
- [ ] TODO: full retest
- [ ] TODO: consolidate OC.Files.FileInfo with OCA.Files.FileInfoModel

**Comment by DeepDiver1975** [comment]
move to 9.0 as discussed

**Comment by PVince81** [comment]
Rebased.

- [ ] fix the many failing unit tests
- [ ] sort out download spinner with "downloadSecret" cookie magic, make it work with Webdav too...

**Comment by PVince81** [comment]
And here we go. Time travel, cherry picking, assembly, beautifing the git history: now this PR is based on davclient.js and also works in IE8 thanks to an extra custom written shim.

I'll update the checkboxes, everything will need to be manually retested.
But first: fixing the unit tests...

**Comment by MorrisJobke** [comment]
> And here we go. Time travel, cherry picking, assembly, beautifing the git history: now this PR is based on davclient.js and also works in IE8 thanks to an extra custom written shim.

You are my super hero of october 2015 :smiley:

**Comment by PVince81** [comment]
- [ ] BUG: public link with password (will push the fix later)

**Comment by PVince81** [comment]
- [ ] BUG: cannot download single file due to "downloadStartSecret" argument being wrongly appended on master, fix is here https://github.com/owncloud/core/pull/20181

**Comment by PVince81** [comment]
Fixed the JS unit tests, they all pass now.

And now we're finally back on our feet and will be able to continue working on the remaining open issues!

**Comment by PVince81** [comment]
- [ ] REQUIRES: https://github.com/owncloud/core/pull/20188 to get access to the sharer's display name as we had it before in ajax/list.php

**Comment by PVince81** [comment]
- [ ] TODO: update bower.json to point to evert's library instead of mine once all patches were merged

**Comment by PVince81** [comment]
- [ ] REQUIRES: davclient.js fix for more complex properties: https://github.com/evert/davclient.js/pull/10

**Comment by PVince81** [comment]
- [ ] TODO: what happens if fileid length > 8 digits ? might need proper webdav API or instance id... (in Webdav oc:fileid is a compound component)

**Comment by Henni** [comment]
The url generation seems to have a problem with port numbers:
image

**Comment by PVince81** [comment]
@Henni nice catch

- [ ] BUG: fix port number handling

**Comment by DeepDiver1975** [comment]
rebased after merge of #20544

**Comment by DeepDiver1975** [comment]
Load order of js files is broken ....

**Comment by PVince81** [comment]
- [ ] REQUIRES: oc:fileid property: https://github.com/owncloud/core/pull/20567

**Comment by PVince81** [comment]
Ok, I'll take care of making it use the oc:fileid field and also updating davclient.js

**Comment by DeepDiver1975** [comment]
@PVince81 I'm currently having a look at the failing phpunit tests

**Comment by PVince81** [comment]
- [ ] TODO: remove "text file" from the new menu in the public page, as the text editor doesn't support public page so far

**Comment by PVince81** [comment]
- [ ] BUG: go to "favorites", rename a folder then directly after that unfavorite from the sidebar. works on master, some info must be missing

**Comment by PVince81** [comment]
This is ready to review and test. There might be a few bugs here and there though.
I already started ticking the boxes. :smile:

**Comment by PVince81** [comment]
- [ ] BUG: upload into sub-subdirectory doesn't display the files until refresh: create "sub1/sub2" and upload files into sub2

**Comment by PVince81** [comment]
- [ ] BUG:  rename file where source folder has been removed => creates a second bogus entry in the list

**Comment by PVince81** [comment]
- [ ] BUG: rename error (ex: invalid chars like "../") recreate a second bogus entry

**Comment by PVince81** [comment]
- [ ] BUG: download spinner doesn't disappear when using Webdav endpoint for single file download (due to missing cookie handling)

**Comment by PVince81** [comment]
- [ ] BUG: owner-display-name doesn't show the correct user for incoming remote shares, raised here https://github.com/owncloud/core/issues/20580

**Comment by PVince81** [comment]
- [ ] TODO: (maybe separately) parse exception messages to be able to show them, requires library support https://github.com/evert/davclient.js/issues/11

**Comment by PVince81** [comment]
Fixed most tests. This PR can already be reviewed and tested.

Next step is to look into the 401 issue, but should not affect the rest of this PR.

**Comment by PVince81** [comment]
Raised separate PR https://github.com/owncloud/core/pull/20609 to fix Ajax+Webdav basic auth dialog, which is more of a general issue once we start using Webdav more often from inside the browser.

**Comment by PVince81** [comment]
The useless "Text file" entry in the "+" menu in the public page will disappear with https://github.com/owncloud/files_texteditor/issues/147 and https://github.com/owncloud/files_texteditor/pull/148

**Comment by PVince81** [comment]
- [ ] BUG: does not work in Firefox ?! Oh noes Firefox, don't let us down :frowning:

**Comment by PVince81** [comment]
- [ ] BUG: broken in IE9 again (invalid document syntax, possible XML mistake)

**Comment by PVince81** [comment]
Turns out that no version of IE supports `document.evaluate()` :rage: 
Back then when it worked in IE9 I was experimenting with another approach, which itself needed an extra workaround for IE8. Now let's see if I can use IE8's workaround for all IE versions...
Else we'll have to move davclient.js away from XPath (the original impl) to something more compatible.

**Comment by PVince81** [comment]
Fortunately the IE8 workaround also works on other IE versions, so I enabled that for now. Raised https://github.com/evert/davclient.js/issues/13 for a future upstream fix, if possible.

**Comment by icewind1991** [comment]
Fixed it to also apply the IE workaround for IE11

**Comment by icewind1991** [comment]
All seems to work on IE11 and Edge

**Comment by PVince81** [comment]
Fantastic! Reverse proxy env also works :smile:

**Comment by MorrisJobke** [comment]
I just tried this with Safari and it looks really good. I did uploads, downloads, moves, renames, creation of folders/files.

* [ ] public link shares requires now the login of the owner (basic auth popup appears)

**Comment by MorrisJobke** [comment]
Beside my comment this looks great :+1:


I'm also fine with merging this and then fix the remaining tasks in new PRs ... otherwise it will be really hard to track the changes.

**Comment by PVince81** [comment]
I increased the app's version, that should hopefully help with the caching.

**Comment by rullzer** [comment]
Did some clicking around.

* Uploaded file
* New file
* New folder
* Deletion
* Moving

All work as expected!
Also did not see any weird stuff when browsing trough the code.

:+1: 
Lets get this in so we actually have some time to find (and fix) issues before the release ;)

**Comment by LukasReschke** [comment]
This looks strange:
```
10:08:39 There was 1 failure:
10:08:39 
10:08:39 1) Test\Files\Storage\SMB::testStat
10:08:39 Failed asserting that 1448204888 is equal to 1448204886 or is less than 1448204886.
10:08:39 
10:08:39 /var/lib/jenkins/jobs/server-master-linux-externals-smb-windows-ext-ci/workspace/database/sqlite/external/smb-windows/label/master/tests/lib/files/storage/storage.php:295
10:08:39 
```

**Comment by DeepDiver1975** [comment]
Clock on the Windows vm is out if sync. Unrelated from my pov

**Comment by MorrisJobke** [comment]
@rullzer @nickvergessen @LukasReschke One more review please :) We need to get this in :)

**Comment by LukasReschke** [comment]
Works :+1:

**Comment by PVince81** [comment]
Great news for a monday morning :smile: 
Thanks guys !

**Comment by LukasReschke** [comment]
:tada: :rocket: :dancer: :beers:

**Comment by felixboehm** [comment]
You rock!

**Comment by jospoortvliet** [comment]
Party time indeed!!! Awesome.

**Pull request merged**